### PR TITLE
.github: dummy make check/API checks for docs-only PRs

### PR DESCRIPTION
**Pull request by djgalloway**
Signed-off-by: David Galloway <dgallowa@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: .github/workflows/docs-only-makecheck.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# The 'make check' check is a required check for all Pull Requests.
+# However, we don't need to run the Jenkins job if a PR only changes
+# documentation. This GitHub Action will create a dummy 'make check'
+# check in lieu of the Jenkins job running and exiting 0 to satisfy the check.
+name: make check
+on:
+  pull_request:
+    paths-ignore:
```
Comment (epuertat) [inline .github/workflows/docs-only-makecheck.yml:8]: I guess you have tested this and it worked, but [Github docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-and-excluding-paths) say you cannot mix paths and paths-ignore in the same event of a workflow.

File: .github/workflows/docs-only-apitests.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      - 'qa/**'
+      - 'selinux/**'
+      - 'src/**'
+      - 'sudoers.d/**'
+      - 'systemd/**'
+      - 'udev/**'
+      - 'ceph.spec.in'
+      - '*.sh'
+      - 'doc_deps.deb.txt'
+      - '*[Mm]ake*'
```
Comment (epuertat) [inline .github/workflows/docs-only-apitests.yml:25]: Can't we define these with a negative pattern? `!/docs/**` and `!*.rst`?

**Review by epuertat — COMMENTED** [review]
I find the approach slightly complicate: we are mixing Jenkins-triggered jobs + Github-triggered actions... I'd stick to one or the other.

Can't we simply use the Checks API within a Jenkins job to indicate that the `make check` passed? Was this related to the lag checking out the Ceph repo? If that's the issue, didn't the depth-1 shallow clone work?

File: .github/workflows/docs-only-makecheck.yml (hunk 1 above)
Comment (djgalloway) [reply .github/workflows/docs-only-makecheck.yml:8]: I'm not sure it's possible to test these without merging them?  I missed that part though.  I guess we'll have to pick one and use `!`

**Comment by djgalloway** [comment]
> I find the approach slightly complicate: we are mixing Jenkins-triggered jobs + Github-triggered actions... I'd stick to one or the other.

I get that but that's kind of the problem.  I suspect GitHub's free resources aren't powerful enough to run our `make check` or `API` jobs.  But at the same time, a docs-only PR shouldn't have to wait for a Jenkins builder to be available just to clone the Ceph repo and list the files changed in a PR.

> Can't we simply use the Checks API within a Jenkins job to indicate that the `make check` passed? Was this related to the lag checking out the Ceph repo? If that's the issue, didn't the depth-1 shallow clone work?

I misrepresented the problem I encountered that led me to want to implement this.  It wasn't that cloning the Ceph repo took a long time.  It was the Jenkins build queue had dozens of jobs lined up and I had to wait a few hours just for the Jenkins job to cloen the repo, see it was docs only, and `exit 0` to make the Check pass.

**Comment by epuertat** [comment]
> I misrepresented the problem I encountered that led me to want to implement this. It wasn't that cloning the Ceph repo took a long time. It was the Jenkins build queue had dozens of jobs lined up and I had to wait a few hours just for the Jenkins job to cloen the repo, see it was docs only, and `exit 0` to make the Check pass.

Aha, ok, now it makes more sense to me. Thanks for the explanation! The more I see the need of creating separate workflows for specific parts of the code base, the more it leads me to think that perhaps we should consider breaking down the monorepo. Would it be a that bad idea to have a docs-only repo?

**Comment by djgalloway** [comment]
> Would it be a that bad idea to have a docs-only repo?

I am not opposed to that.  Maybe a topic for next CLT call?

**Comment by djgalloway** [comment]
@epuertat So since the idea of moving docs to a separate repo was emphatically shot down, I'd love to work together on fixing out the `paths` regex.

I'm actually waiting on https://jenkins.ceph.com/job/ceph-api/41099/console as we speak due to some bug either in Jenkins or the Artifact Archiving plugin to fail or pass or whatever it's going to do in order to merge https://github.com/ceph/ceph/pull/47219.  (Just as an example of why this PR is needed)

**Comment by epuertat** [comment]
> @epuertat So since the idea of moving docs to a separate repo was emphatically shot down, I'd love to work together on fixing out the `paths` regex.
> 
> I'm actually waiting on [jenkins.ceph.com/job/ceph-api/41099/console](https://jenkins.ceph.com/job/ceph-api/41099/console) as we speak due to some bug either in Jenkins or the Artifact Archiving plugin to fail or pass or whatever it's going to do in order to merge #47219. (Just as an example of why this PR is needed)

Understood. Please let me know when those tests passed and I'll review again!

**Review by epuertat — APPROVED** [review]
Looks good from code perspective. Do you have sample runs of this?

**Comment by djgalloway** [comment]
I've been trying to test it using https://github.com/nektos/act/ but it always says the "PR" is a docs-only PR.

I'm not sure how else to test other than merging and opening a docs-only PR and a not-docs-only PR.

**Comment by epuertat** [comment]
> I've been trying to test it using [nektos/act](https://github.com/nektos/act/) but it always says the "PR" is a docs-only PR.
> 
> I'm not sure how else to test other than merging and opening a docs-only PR and a not-docs-only PR.

I sometimes test Github actions from my forked Ceph repo.

**Pull request closed without merging**